### PR TITLE
TYP: stubs for reshape, ops, ops_dispatch, hashing

### PR DESCRIPTION
**Pull request by jbrockmendel**
manually curated, extra specificity in comments it is supported directly

### CODE REVIEW

File: pandas/_libs/ops_dispatch.pyi [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import numpy as np
+
+REVERSED_NAMES: dict[str, str]
+UFUNC_ALIASES: dict[str, str]
+DISPATCHED_UFUNCS: set[str]
```
Comment (simonjayhawkins) [inline pandas/_libs/ops_dispatch.pyi:5]: These are not accessed from python code.

so I don't think is covered by https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#what-to-include

and from https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#conventions

> The general rule is that they should be as concise as possible.

so I think we should perhaps not include these.

once we activate `ignore_missing_imports = False`, we'll know if we missed anything.

---

side note: for the stub files we could consider activating more of the strictness flags for incomplete function annotations and dynamic typing etc.

File: pandas/_libs/ops_dispatch.pyi [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+def maybe_dispatch_ufunc_to_dunder_op(
+    self, ufunc: np.ufunc, method: str, *inputs, **kwargs
```
Comment (simonjayhawkins) [inline pandas/_libs/ops_dispatch.pyi:8]: maybe type self, since function is not a method

File: pandas/_libs/ops.pyi [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+def vec_binop(
+    left: np.ndarray,   # object[:]
+    right: np.ndarray,  # object[:]
+    op: binop,         # binary operator
+) -> np.ndarray: ...
+
+
+def maybe_convert_bool(
+    arr: np.ndarray,  # np.ndarray[object]
+    true_values=None,
```
Comment (simonjayhawkins) [inline pandas/_libs/ops.pyi:41]:
```suggestion
    true_values=...,
```

File: pandas/_libs/ops.pyi [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    false_values=None
```
Comment (simonjayhawkins) [inline pandas/_libs/ops.pyi:42]:
```suggestion
    false_values=...
```

File: pandas/_libs/ops.pyi [hunk 5]
```diff
@@ -0,0 +1,43 @@
+from typing import (
+    Any,
+    Callable,
+)
+
+import numpy as np
+
+binop = Callable[[Any, Any], Any]
+bool_op = Callable[[Any, Any], bool]
```
Comment (simonjayhawkins) [inline pandas/_libs/ops.pyi:9]: from https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#conventions

> Type variables and aliases you introduce purely for legibility reasons should be prefixed with an underscore to make it obvious to the reader they are not part of the stubbed API.

we normally capitalize as well

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @jbrockmendel generally lgtm

**Comment by simonjayhawkins** [comment]
> manually curated, extra specificity in comments it is supported directly

IIUC The pyright stubs are mainly for the public api. for our internal consistency checking , pandas._libs includes functions that we use internally, so manually curated is probably best for now.

**Comment by jbrockmendel** [comment]
Will update per suggestions.

> IIUC The pyright stubs are mainly for the public api. for our internal consistency checking , pandas._libs includes functions that we use internally, so manually curated is probably best for now.

Following the dev call last month (two months ago?) about stubs I looked over pyright's stubs for _libs and found them not-totally-accurate (mostly Timestamp/Timedelta IIRC), so I plan to do all of these manually.

File: pandas/_libs/ops_dispatch.pyi (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/ops_dispatch.pyi:8]: will take some tracking down, since it is called via arraylike.array_ufunc, which doesnt type self

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**